fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -253,51 +254,64 @@ private boolean shouldClearExistingIndexEntries(@Nonnull IndexState indexState,
-                                    throw new RecordCoreException("This index was partly built by another method");
-                                }
-                                return null;
-                            }); },
-                false,
-                additionalLogMessageKeyValues);
+                    .thenApply( bytes -> {
+                        if (bytes == null || forceStampOverwrite) {
+                            transaction.set(stampKey, indexingTypeStamp.toByteArray());
+                        } else {
+                            IndexBuildProto.IndexBuildIndexingStamp savedStamp ;
+                            try {
+                                savedStamp = IndexBuildProto.IndexBuildIndexingStamp.parseFrom(bytes);
+                            } catch (InvalidProtocolBufferException ex) {
+                                savedStamp = null;
+                            }
+                            if ( savedStamp == null || ! matchingIndexingTypeStamp(savedStamp)) {
+                                throw new RecordCoreException("This index was partly built by another method",
+                                        LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
+                                        LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
+                                        LogMessageKeys.INDEXER_ID, common.getUuid(),
+                                        LogMessageKeys.EXPECTED_TYPE, indexingTypeStamp,
+                                        LogMessageKeys.ACTUAL_TYPE, savedStamp);
+                            }
+                        }
+                        return null;
+                    });
+        } ));
     }
 
     @Nonnull
-    abstract byte[] getTypeStamp();
+    abstract IndexBuildProto.IndexBuildIndexingStamp getIndexingTypeStamp();
 
     @Nonnull
-    abstract boolean matchingTypeStamp(byte[] stamp);
+    abstract boolean matchingIndexingTypeStamp(@Nonnull final IndexBuildProto.IndexBuildIndexingStamp stamp);
 
     abstract CompletableFuture<Void> buildIndexByEntityAsync();
 
     // Helpers for implementing modules. Some of them are public to support unit-testing.
-    protected boolean shouldLogBuildProgress() {
+    protected void maybeLogBuildProgress(SubspaceProvider subspaceProvider, List<Object> additionalLogMessageKeyValues) {
+        if (LOGGER.isInfoEnabled() && shouldLogBuildProgress()) {
+            final Index index = common.getIndex();
+            LOGGER.info(KeyValueLogMessage.of("Built Range",
+                    LogMessageKeys.INDEX_NAME, index.getName(),
+                    LogMessageKeys.INDEX_VERSION, index.getLastModifiedVersion(),
+                    subspaceProvider.logKey(), subspaceProvider,
+                    LogMessageKeys.RECORDS_SCANNED, common.getTotalRecordsScanned().get()),
+                    LogMessageKeys.INDEXER_ID, common.getUuid(),
+                    additionalLogMessageKeyValues);
```

Comment:
If you go down this route, I think you want something like:

```suggestion
            LOGGER.info(KeyValueLogMessage.build("Built Range",
                   LogMessageKeys.INDEX_NAME, index.getName(),
                   LogMessageKeys.INDEX_VERSION, index.getLastModifiedVersion(),
                   subspaceProvider.logKey(), subspaceProvider,
                   LogMessageKeys.RECORDS_SCANNED, common.getTotalRecordsScanned().get(),
                   LogMessageKeys.INDEXER_ID, common.getUuid())
                   .addKeysAndValues(additionalLogMessageKeyValues)
                   .toString());
```

(Alternatively, breaking it up into multiple statements might make sense instead of using the fluent syntax might make sense.) The new code adds the common keys to the message and then adds the additional keys to the end.

I also expect the code as written to throw `IllegalArgumentException("keys and values don't match")` but for a second bug where an extra `)` was added after the `RECORDS_SCANNED` pair.